### ISSUE TITLE
support extracting filesystem from kyocera printer firmware

### ISSUE THREAD

**Issue by Yannik** [devttys0/binwalk#178]
Hi,

I am trying to extract a kyocera printer firmware. 

Running `binwalk -e` on the firmware file returns the following results:
```
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
15917         0x3E2D          Zlib compressed data, default compression
13812163      0xD2C1C3        mcrypt 2.5 encrypted data, algorithm: "", keysize: 27369 bytes, mode: "M",
26231768      0x19043D8       LZMA compressed data, properties: 0x63, dictionary size: 16777216 bytes, uncompressed size: 2883884 bytes
26632662      0x19661D6       MySQL ISAM compressed data file Version 1
28384990      0x1B11EDE       YAFFS file system, big endian
```

binwalk extracts the a file `3E2D.zlib` from 0x3E2D to the end of the file and uncompresses it to a file `3E2D`. `strings 3E2D` shows many familiar strings belonging to ELFs and other files (static assets for the webinterface).

`binwalk -e 3E2D` results in > 10GB of .7z files
output:
```
30629505      0x1D35E81       LZMA compressed data, properties: 0x76, dictionary size: 16777216 bytes, uncompressed size: 803209216 bytes
30629825      0x1D35FC1       LZMA compressed data, properties: 0xB6, dictionary size: 16777216 bytes, uncompressed size: 266403840 bytes
30631305      0x1D36589       LZMA compressed data, properties: 0x7E, dictionary size: 16777216 bytes, uncompressed size: 467992576 bytes
....
```
the uncompressed size is unrealistic so these must be false positives.

To exclude these I run `binwalk -e -x 'lzma compressed data' 3E2D`.
This shows a `VxWorks WIND kernel version "2.9"`, various ELFs, Unix Paths and other files, including a 
```
31559244      0x1E18E4C       Linux EXT filesystem, rev 0.0, ext2 filesystem data (mounted or unclean), UUID=00000000-0000-0003-cce2-000234bb34bb
```
entry quite in the beginning. The extracted file `1E18E4C.ext` contains many interesting strings belonging to cgi ELFs and static web assets.

Any ideas on how to properly extract this filesystem / what kind of filesystem this is?

The firmware zip can be downloaded here: 
https://mega.nz/#!bxg1QLrC!SnEZm4xUHeGmpRVP9CuzYWtXYVHS3PH7FPFv4K7A9d0

Edit: This probably isn't an ext2fs but something in the direction of Fat16. 

Maybe this helps:
```
^@^@^@^A`Aÿ^@^@<94>\¾wÀF¼wÿÿÿÿ^@^@^@^@^@^@^@^@^@^@^@^V^@^@^A`^@ÐîÄ^A^A.^@^@ÐÇø^A^B..^@ÐðH¶^Kversion.htm^@^@^@^@ÐóL¶^Gtop.htm^@^@^@^@ÐúÔ¶        index.htm^@^@Ñ^H<8c>¶^Kfavicon.ico^@^@^@^@Ñ^L0ÿ^Cadv^@^@^@^@ØØ¬ÿ^Ebasic^@^@Û|(ÿ^Ccss^@^@^@^@ÛÍ<9c>ÿ^Cdeu^@^@^@^@ÛÒ ÿ^Cdut^@^@^@^@Û×¤ÿ^Ceng^@^@^@^@ÛÜ¨ÿ^Eerror^@^@Ûátÿ^Cesp^@^@^@^@Ûæxÿ^Heventlog^@^@^@Ü^F<98>ÿ^Cfra^@^@^@^@Ü^K<9c>ÿ^Fimages^@Ü¤Äÿ^Cita^@^@^@^@Ü©Èÿ^Cjob^@^@^@^@ÝF8ÿ^Cjpn^@^@^@^@ÝK<ÿ^Bjs^@ßH¸ÿ^Dmail^@^@^@ßm´ÿ^Dopt1^@^@^@ßr¼ÿ^Gprinter^@^@^@^@à^G^\ÿ^Crus^@^@^@^@à^L ÿ^Estart^@^@ã6 ÿ^Gstrings^@^@^@^@^@^@^@^@^@^BÞ<81>¶^@^@<94>\¾wÀF¼wÿÿÿÿ^@^@^@^@^@^@^@^@^@^@^@^V^@^@^Bà<html dir="ltr">^M
<head>^M
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">^M
<meta http-equiv="Content-Style-Type" content="text/css">^M
<title></title>^M
<link rel="stylesheet" href="css/main.css">^M
</head>^M
<body style="backGround:#fff">^M
<div id="title"><h1>Kyocera Command Center</h1>^M
<br>^M
<br>^M
<b>Target :</b> 2GR^M
<br><br>^M
<b>Template Version :</b> 006053001^M
<br><br>^M
<b>Specification Version :</b> Not Modified^M
</div>^M
<div id="content">^M
<B>ä¸»ã<81>ªæ<9b>´æ<96>°å<86><85>å®¹</B>^M
<UL style="margin-top:3px;margin-bottom:5px">^M
<LI>Local Navigationã<81><8c>IE5.5ã<81>§æ­£å¸¸ã<81>«å<8b><95>ä½<9c>ã<81><97>ã<81>ªã<81><84>ã<81>®ã<82><92>ä¿®æ­£</LI>^M
</UL>^M
<B>å<88>¶é<99><90>äº<8b>é <85></B>^M
<UL style="margin-top:3px;margin-bottom:5px">^M
<LI>ã<81>ªã<81><97></LI>^M
</UL>^M
</body>^M
</html>^M
^@^@^@^@^Gb<81>¶^@^@<94>\¾wÀF¼wÿÿÿÿ^@^@^@^@^@^@^@^@^@^@^@^V^@^@^Gd<!------------------------------------------------------------------------->^M
<!-- Kyocera Command Center                                              -->^M
<!-- for Project code 2GR                                                -->^M
<!--                                                                     -->^M
<!-- Copyright (C) 2005 KYOCERA MITA Corporation                         -->^M
<!------------------------------------------------------------------------->^M
<!-- File Name : top.htm                                                 -->^M
<!-- Category  : Topframe temporary file                                 -->^M
<!-- Kind      : PageSource                                              -->^M
<!------------------------------------------------------------------------->^M
<!-- Original    :                                                       -->^M
<!-- Modify      :                                                       -->^M
<!-- Last Update : Masato Mine                            Date: 7.14.2005-->^M
<!------------------------------------------------------------------------->^M
<!-- Note:                                                               -->^M
<!--   Please use the charactor-code "UTF-8" when you edit this source.  -->^M
<!--   The greater part of comments are written in Japanese.             -->^M
<!--                                                                     -->^M
<!-- ã<82>³ã<83>¡ã<83>³ã<83><88>ã<81>¯è<8b>±èª<9e>ã<81><8c>æ¨<99>æº<96>ã<80><82>æ<97>¥æ<9c>¬èª<9e>ã<81>¯æ<89><80>ã<81>«ã<82><88>ã<82><8a>å<98><98>ã<81><8c>ã<81><82>ã<82><8b>ã<81><8b>ã<82><82>w -->^M
<html dir="ltr">^M
<head>^M
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">^M
<meta http-equiv="Content-Style-Type" content="text/css">^M
<meta http-equiv="Content-Script-Type" content="text/javascript">^M
<title></title>^M
</head>^M
<body bgcolor="#000000">^M
<form>^M
<script language="JavaScript">^M
</script>^M
<input type="hidden" name="FileName" value="">^M
</form>^M
</body>^M
</html>^M
^@^@^@^@^M<94><81>¶^@^@<94>\¾wÀF¼wÿÿÿÿ^@^@^@^@^@^@^@^@^@^@^@^V^@^@^M<94><!------------------------------------------------------------------------->^M
<!-- Kyocera Command Center                                              -->^M
<!-- for Project code 2GR                                                -->^M
<!--                                                                     -->^M
<!-- Copyright (C) 2005 KYOCERA MITA Corporation                         -->^M
<!------------------------------------------------------------------------->^M
<!-- File Name : index.htm                                               -->^M
```

First there is a listing of files and directories
(I found this list of directories in the http server config file:
```
/start/
/basic/
/printer/
/scanner/
/job/
/box/
/adv/
/deu/index.htm
/dut/index.htm
/eng/index.htm
/esp/index.htm
/fra/index.htm
/ita/index.htm
/jpn/index.htm
/rus/index.htm
/opt1/index.htm
/fax/
/mail/
/top.htm
/index.htm
/eventlog
```

After that index table, the file data of the listed files (version.htm, top.htm, index.htm) seem to follow.

**Comment by devttys0** [comment]
I don't get the LZMA false positives that you do; your system probably doesn't have the `python-lzma` module installed which binwalk uses to further validate LZMA signatures.

All the file system results are almost certainly false positives. As binwalk suggests, this firmware is VxWorks based, not Linux. Although VxWorks supports some file systems, my experience with it is that there are typically no file systems (your suggestion of FAT16 is possible), at least not as most people would think of them. There are often just some basic data structures with file names, pointers to the file data, and a size field to specify the file size. These are usually used for storing HTML data for the web interface, as the file paths and names that you posted suggest.

I did notice an unusual number of ELF files for a VxWorks image, but they are valid ELFs complete with symbol tables, and looking at the strings and function names they are VxWorks specific. AFAICT, these are VxWorks executables, probably intended to be executed by a user from some command line interface (UART, telnet, ssh, etc).

The main code though is the VxWorks kernel, which appears to be the decompressed `3E2D` file. Both `3E2D` and the main firmware image contain big endian PowerPC code (according to `binwalk -A`), and may or may not contain symbol tables or other helpful information to aid in disassembly  / reverse engineering. These are typically blobs of code and data mashed together, so you'll need to identify where code ends and data starts (`binwalk -A` can help point you in the right direction), as well as the code load address in order to get a readable disassembly.

**Comment by Yannik** [comment]
>I don't get the LZMA false positives that you do; your system probably doesn't have the python-lzma module installed which binwalk uses to further validate LZMA signatures.

I just installed python-backports-lzma and was able to `from backports import lzma`. However, binwalk still shows the false positive lzma matches. Using python3 it doesn't. Maybe this is a bug that should be fixed.

As the device supports a huge number of protocols (SNMPv3, Netware, AppleTalk, NetBEUI, Raw Printing, IPP, IPPS, HTTP/HTTPS, FTP, SNMP, POP3, SMTP, LPD, WSD, IPSec, 802.1x), a larger count of ELFs seems reasonable to me. I will maybe investigate into the processor model to find information about the load address later on, the symbol tables being intact is a positive find. Regularly, there is no cli access to users. 
Is it quite certain that everything is baked into the kernel? It would be a bit easier to investigate separated files/modules if it was possible to extract a usable filesystem.

**Comment by devttys0** [comment]
>I just installed python-backports-lzma...binwalk still shows the false positive lzma matches. Using python3 it doesn't. Maybe this is a bug that should be fixed.

Binwalk expects the module to be named `lzma`, not `backports.lzma`. Most Linux distros have a `python-lzma` package for Python2; it is a standard module in Python3, which is why you don't get the LZMA false positives in Python3. I've updated the master branch to check for `backports.lzma`, so that should be fixed now.

>a larger count of ELFs seems reasonable to me.

Well, bear in mind when I say "larger" I mean there are 5 or 6 ELF files. Usually with VxWorks I've only seen one ELF file, which was for the decompression stub, so it was not part of the main VxWorks OS.

> the symbol tables being intact is a positive find

The symbol tables for the ELF files are intact, but I didn't immediately see a symbol table for the main OS.

>Regularly, there is no cli access to users.

I did a very brief disassembly of some of the ELF files, and some have command line usage strings and `getopt` style strings in them, so they do take command line options. Could just be for developer debug access via a serial port or something though.

>Is it quite certain that everything is baked into the kernel? It would be a bit easier to investigate separated files/modules if it was possible to extract a usable filesystem.

It would be a lot easier, but unfortunately that's usually how Real Time Operating Systems (like VxWorks) work: the OS is typically one big kernel running in a flat memory space (there is no user space or user applications). With that said, some RTOS's support user space protected code execution, but I'm not sure how that would work in VxWorks.
